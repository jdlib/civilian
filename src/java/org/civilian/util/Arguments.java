/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.util;


import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;


/**
 * An iterator for command line arguments.
 * Any argument prefixed by '@' is treated as argument file: Its content
 * is read, interpreted as arguments and inserted into the argument list. 
 */
public class Arguments
{
	/**
	 * Creates a Arguments object.
	 * @param args the string arguments to the main method
	 */
	public Arguments(String... args)
	{
    	this(null, args);
    }
    
    
    /**
	 * Creates a Arguments object.
     * @param loader if not null then it is used to open argument files
     * @param args the arguments
     */
    public Arguments(ResourceLoader loader, String... args)
    {
    	loader_	= loader;
    	addArgs(args, null);
	}
   
    
    private ArrayList<String> addArgs(String[] args, ArrayList<String> files)
    {
    	int len = args == null ? 0 : args.length;
    	
    	if (args_ == null)
    		args_ = new ArrayList<>(len);
    	else
    		args_.ensureCapacity(args_.size() + len);
    			
		for (int i=0; i<len; i++)
		{
			String arg = args[i];
			
			// ignore null and empty args
			if ((arg != null) && (arg.length() > 0))
			{
				if ((arg.charAt(0) == '@') && (arg.length() > 1))
					files = insertArgumentFile(arg.substring(1), files);
				else 
					args_.add(arg);
			}
		}
		
		return files;
    }

	
	private ArrayList<String> insertArgumentFile(String file, ArrayList<String> files)
	{
		if (files == null)
			files = new ArrayList<>();
		if (files.contains(file))
			throw new IllegalArgumentException("cyclic inclusion of argument file: " + file);
		files.add(file);
		
		try
		{
			try(InputStream in = loader_ == null ?  
				new FileInputStream(file) :
				loader_.required().getResourceAsStream(file))
			{
	            String args[] = readArguments(in);
	            files = addArgs(args, files);
	            return files;
			}
		}
		catch(RuntimeException e)
		{
			throw e;
		}
		catch(Exception e)
		{
			throw new IllegalArgumentException("cannot read argument file " + file, e);
		}
	}


    private String[] readArguments(InputStream in) throws IOException
    {
    	BufferedReader reader = new BufferedReader(new InputStreamReader(in, "UTF-8"));
    	String line;
    	ArrayList<String> args = new ArrayList<>();
    	while((line = reader.readLine()) != null)
    	{
    		line = line.trim();
    		if (!line.startsWith("#"))
    		{
	    		StringTokenizer st = new StringTokenizer(line);
	    		while(st.hasMoreTokens())
	    			args.add(st.nextToken());
    		}
    	}
    	return args.toArray(new String[args.size()]);
    }
    
 
    /**
	 * @return the current argument.
	 */
	public String get()
	{
		return hasMore() ? args_.get(pos_) : null;
	}

	
	/**
	 * @return the next argument string and proceeds to the next argument.
	 */
	public String next()
	{
		return next(null);
	}


	/**
	 * @return the next argument string or throws an IllegalArgumentException if there
	 * are no more arguments left.
	 * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
	 */
	public String next(String what)
	{
		if (!hasMore())
		{
			String message = what != null ? "missing argument: " + what : "no arguments left"; 
			throw new IllegalArgumentException(message);
		}
		return args_.get(pos_++);
	}


    /**
     * @return the next argument string as int.
 	 * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
     * @throws IllegalArgumentException if there are no more arguments left.
     */
    public int nextInt(String what)
    {
        String s = next(what);
        try
		{
        	return Integer.parseInt(s);
		}
        catch(NumberFormatException e)
		{
    		throw createConvertException(what, "'" + s + "' is not a integer", e);
		}
    } 
    
    
    /**
     * @return the next argument string as boolean.
	 * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
     * @throws IllegalArgumentException if there are no more arguments left.
     */
    public boolean nextBoolean(String what)
    {
        String s = next(what);
       	return Boolean.parseBoolean(s);
    } 

    
    /**
     * @return the next argument string as File.
 	 * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
     * @throws IllegalArgumentException if there are no more arguments left.
     */
    public File nextFile(String what)
    {
    	return nextFile(what, null);
    }
    
    
    /**
     * @return the next argument string as File.
     * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
     * @param fileType the expected file type.
     * @throws IllegalArgumentException if there are no more arguments left or the file
     * 		does not match the expected file type
     */
    public File nextFile(String what, FileType fileType)
    {
        File file = new File(next(what));
        if (fileType != null)
        	fileType.check(file, what);
        return file;
    } 

    
    /**
     * Interprets the next argument string as class name and returns a
     * new instance of that class.
     * @param what describes the parameter. Used when a IllegalArgumentException is thrown.
     * @param superClass a super class of the class.
     * @param <T> the superclass type
     * @return the new instance
     */
    public <T> T nextObject(String what, Class<T> superClass)
    {
    	String className = next(what);
    	try
    	{
    		return ClassUtil.createObject(className, superClass, null);
    	}
    	catch(Exception e)
    	{
    		throw createConvertException(what, e.getMessage(), e);
    	}
    } 

   
    /**
	 * @param s a string 
	 * @return if the next argument matches the string. If true, 
	 * the string is consumed.
	 */
	public boolean consume(String s)
	{
		if (hasMore() && args_.get(pos_).equals(s))
		{
			pos_++;
			return true;
		}
		else
			return false;
	}


	/**
	 * @param s a string
	 * Replaces the current argument with the string.
	 */
	public void replace(String s)
	{
		if (hasMore())
			args_.set(pos_, s);
	}
	
	
	/**
	 * @param s a string
	 * @return if the next argument starts with the given string.
	 */
	public boolean startsWith(String s)
	{
		return hasMore() && args_.get(pos_).startsWith(s);
	}


	/**
	 * @return if there are more arguments left.
	 */
	public boolean hasMore()
	{
		return pos_ < args_.size();
	}


	/**
	 * @param count the count
	 * @return if there are at least count more arguments.
	 */
	public boolean hasMore(int count)
	{
		return pos_ + count <= args_.size();
	}
	
	
	/**
	 * @return the remaining arguments.
	 */
	public List<String> getRestArgs()
	{
		return args_.subList(pos_, args_.size());
	}
	
	
	private IllegalArgumentException createConvertException(String what, String message, Exception cause)
	{
		if (what != null)
			message = "reading " + what + ": " + message;
		return new IllegalArgumentException(message, cause);
	}


	private ArrayList<String> args_;
	private int pos_;
    private ResourceLoader loader_;
}
