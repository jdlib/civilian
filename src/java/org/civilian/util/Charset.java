/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.util;


/**
 * Charset is a small helper class  to determine if a character
 * is printable in a certain charset. (E.g. IS=-8859-1 can only
 * print characters &lt;= 0xff). We use this class
 * in HTML output to determine if we need to print a character reference
 * instead of the character itself.
 * This is just used as a conservative heuristic so we do not intent
 * comprehensive coverage of charsets (We wish we could get 
 * the information from java.nio.charset.Charset).
 */
public abstract class Charset
{
	public static final Charset UNRESTRICTED 	= new UnrestrictedCharset();
	public static final Charset SEVEN_BIT		= new SimpleCharset(0x007f);
	public static final Charset EIGHT_BIT 		= new SimpleCharset(0x00ff);
	
	
	/**
	 * Returns a Charset for the charset name. If not known a eight-bit charset
	 * is returned.
	 * @param charsetName a charset name like UTF-8, ISO-8859-1
	 */
	public static Charset getCharset(String charsetName)
	{
		if (charsetName == null)
			return EIGHT_BIT; // don't bother 
		
		charsetName = charsetName.toUpperCase();
		if (charsetName.startsWith("UTF-"))
			return UNRESTRICTED;
		else if (startsWith(charsetName, "ISO-8859-"))
			return EIGHT_BIT; // actually ISO-8859-2 could print more 
		else if (startsWith(charsetName, "EBCDIC-CP-"))
			return EIGHT_BIT;
		else if (startsWith(charsetName, "EUC-"))
			return UNRESTRICTED;
		else if (startsWith(charsetName, "ASCII"))
			return SEVEN_BIT;
		else
			return EIGHT_BIT;
	}
	
	
	private static boolean startsWith(String name, String prefix)
	{
		return name.regionMatches(true /*ignore-case*/, 0, prefix, 0, prefix.length());
	}
	
	
	/**
	 * Tests if the charset can print the character.
	 */
	public abstract boolean isPrintable(char c);
}


class SimpleCharset extends Charset
{
	public SimpleCharset(int lastPrintable)
	{
		lastPrintable_ = lastPrintable;
	}
	
	
	@Override
	public boolean isPrintable(char c)
	{
		return c <= lastPrintable_;
	}
	
	
	private int lastPrintable_;
}



class UnrestrictedCharset extends Charset
{
	@Override public boolean isPrintable(char c)
	{
        // in unicode not entirely correct for one half of surrogate pairs
		// (c<0xD800) || (c>0xDBFF)
        return true;
	}
}