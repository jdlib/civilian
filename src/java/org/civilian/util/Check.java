/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.util;


/**
 * Check defines helper methods for argument checking.
 */
public class Check
{
	/**
	 * Checks that an object is not null.
	 * @param object the object
	 * @param what describes the object.
	 * @return the object
	 * @exception IllegalArgumentException if the object is null.
	 * @param <T> the class type
	 */
	public static <T> T notNull(T object, String what)
	{
		if (object == null)
			throw new IllegalArgumentException(what + " is null");
		return object;
	}
	
	
	/**
	 * Checks that an object is null.
	 * @param object the object
	 * @param what describes the object.
	 * @exception IllegalArgumentException if the object is null.
	 */
	public static void isNull(Object object, String what)
	{
		if (object != null)
			throw new IllegalArgumentException(what + " is already set");
	}


	/**
	 * Checks that an array is not null and has length &gt; 0.
	 * @param array the array
	 * @return the array
	 * @param what describes the array.
	 * @exception IllegalArgumentException if the array is empty.
	 * @param <T> the class type
	 */
	public static <T> T[] notEmpty(T[] array, String what)
	{
		notNull(array, what);
		if (array.length == 0)
			throw new IllegalArgumentException(what + " empty");
		return array;
	}


	/**
	 * Checks that a CharSequence is not null and has length &gt; 0.
	 * @param s the CharSequence
	 * @param what describes the string
	 * @param <T> the CharSequence type
	 * @return the CharSequence
	 * @exception IllegalArgumentException if the CharSequence is empty.
	 */
	public static <T extends CharSequence> T notEmpty(T s, String what)
	{
		notNull(s, what);
		if (s.length() == 0)
			throw new IllegalArgumentException(what + " empty");
		return s;
	}


	/**
	 * Checks that a double is between [min, max].
	 * @param v the double
	 * @param min the minimum
	 * @param max the maximum
	 * @param what describes the value
	 * @return the value
	 * @exception IllegalArgumentException if the value is outside the range.
	 */
	public static double between(double v, double min, double max, String what)
	{
		if ((v < min) || (v > max))
			throw new IllegalArgumentException(what + " must be between " + min + " and " + max + ": " + v);
		return v;
	}


	/**
	 * Checks that a int is between [min, max].
	 * @param v the values
	 * @param min the minimum
	 * @param max the maximum
	 * @param what describes the value
	 * @return the value
	 * @exception IllegalArgumentException if the value is outside the range.
	 */
	public static int between(int v, int min, int max, String what)
	{
		if ((v < min) || (v > max))
			throw new IllegalArgumentException(what + " must be between " + min + " and " + max + ": " + v);
		return v;
	}

	
	/**
	 * Checks that a int value is &gt;= another value.
	 * @param n the value
	 * @param min the minimum
	 * @param what describes the value
	 * @return the value
	 * @exception IllegalArgumentException if the value is outside the range.
	 */
	public static int greaterEquals(int n, int min, String what)
	{
		if (n < min)
			throw new IllegalArgumentException(what + " must >= " + min + ", but is " + n);
		return n;
	}
	

	/**
	 * Checks that the first class is a superclass of the second class.
	 * @param superClass the super class candidate
	 * @param derivedClass the derived class candidate
	 */
	public static void isSuperclassOf(Class<?> superClass, Class<?> derivedClass)
	{
		if (!superClass.isAssignableFrom(derivedClass))
			throw new IllegalArgumentException("class '" + derivedClass.getName() + "' is not derived from '" + superClass.getName() + "'");
	}


	public static <T> T isA(Object object, Class<T> type)
	{
		return isA(object, type, null);
	}


	/**
	 * Checks that an object is an instance of a class.
	 * @param object the object
	 * @param type a class
	 * @param what describes the object.
	 * @return the object
	 * @exception IllegalArgumentException if the object is null.
	 * @param <T> the class type
	 */
	@SuppressWarnings("unchecked")
	public static <T> T isA(Object object, Class<T> type, String what)
	{
		notNull(object, what);
		if (!type.isAssignableFrom(object.getClass()))
			isAFailed(object, type, what);
		return (T)object;
	}


	private static void isAFailed(Object object, Class<?> type, String what)
	{
		String cname = object.getClass().getName();
		String descr = what == null ? cname : what + " (" + cname + ')';
		if (!cname.equals(type.getName()))
			throw new IllegalArgumentException(descr + " is not a " + type.getName());
		else
			throw new IllegalStateException(descr + " was loaded by " + object.getClass().getClassLoader() + " and not by " + type.getClassLoader());
	}
}
