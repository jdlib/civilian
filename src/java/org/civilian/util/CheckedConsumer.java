package org.civilian.util;


@FunctionalInterface
public interface CheckedConsumer<T,E extends Exception> 
{
	public void accept(T value) throws E;
}
