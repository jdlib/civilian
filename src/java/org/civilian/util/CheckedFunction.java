package org.civilian.util;


import java.util.function.Function;


/**
 * Represents a function that accepts one argument and produces a result.
 * Similar to {@link Function} but allows to throw checked exceptions.
 * @param <T> the type of the input to the function
 * @param <R> the type of the result of the function
 * @param <E> the type of the Exception
 */
@FunctionalInterface
public interface CheckedFunction<T, R, E extends Exception> 
{
    /**
     * Applies this function to the given argument.
     *
     * @param t the function argument
     * @return the function result
     */
    R apply(T t) throws E;
}
