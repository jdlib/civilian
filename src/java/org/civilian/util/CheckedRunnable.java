package org.civilian.util;


@FunctionalInterface
public interface CheckedRunnable<E extends Exception> 
{
	public void run() throws E;
	
	public default Runnable unchecked() 
	{
		return () -> {
			try
			{
				run();
			}
			catch (RuntimeException e)
			{
				throw e;
			}
			catch (Exception e)
			{
				throw new IllegalStateException("error when running " + this, e);
			}
		};
	}
}
