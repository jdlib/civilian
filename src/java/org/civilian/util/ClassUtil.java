/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.util;


import java.lang.annotation.Annotation;


/**
 * A collection of class, annotation and Enum related utility methods.
 */
public abstract class ClassUtil
{
	/**
	 * Tests if two objects are equal.
	 */
	public static boolean equals(Object s1, Object s2)
	{
		return s1 == null ? s2 == null : s1.equals(s2);
	}
	

	/**
	 * Returns the package part of the class name (without a trailing dot).
	 * Alternatively you could use Class.getPackage.getName(), but
	 * since Class.getPackage could return null, this method is safer.
	 */
	public static String getPackageName(Class<?> c)
	{
		return getPackageName(c != null ? c.getName() : null);
	}


	/**
	 * Returns the package part of the given class name (without a trailing dot).
	 * 
	 */
	public static String getPackageName(String className)
	{
		if (className == null)
			return null;
		else
		{	
			int p = className.lastIndexOf('.');
			return (p == -1) ? null : className.substring(0, p);
		}
	}
	
	
	/**
	 * Removes the package from the class name and returns the rest.
	 * In case of an inner class, this will return
	 * the inner class name, prefixed by all outer class names. 
	 */
	public static String cutPackageName(String className)
	{
		if (className == null)
			return null;
		else
		{
			int p = className.lastIndexOf('.');
			return (p == -1) ? className : className.substring(p + 1);
		}
	}
	
	
	/**
	 * Tests if the object is an instance of the given class.
	 */
	public static <T> boolean isA(Object object, Class<T> c)
	{
		return (object != null) && (c != null) && c.isAssignableFrom(object.getClass());
	}


	/**
	 * Creates an object instance for a given class name.
	 * @param className the name of the class
	 * @param superClass a superclass of the class. The method checks that the class 
	 * 		is derived from the superclass. 
	 * @throws ClassNotFoundException if the class was not found 
	 * @throws IllegalArgumentException if the class is not derived from the given superClass. 
	 * @throws InstantiationException if the instance could not be created 
	 * @throws IllegalAccessException if the instance could not be created 
	 */
	@SuppressWarnings("deprecation")
	public static <T> T createObject(String className, Class<T> superClass, ClassLoader loader) 
		throws ClassNotFoundException, InstantiationException, IllegalAccessException
	{
		Class<? extends T> c = getClass(className, superClass, loader);
		return c.newInstance();
	}
	
	
	/**
	 * Returns a class for a given class name.
	 * @param className the name of the class
	 * @param superClass a superclass of the class. The method checks that the class 
	 * 		is derived from the superclass. 
	 * @throws ClassNotFoundException if the class was not found 
	 * @throws IllegalArgumentException if the class is not derived from the given superClass. 
	 */
	@SuppressWarnings("unchecked")
	public static <T> Class<? extends T> getClass(String className, Class<T> superClass, ClassLoader loader)
		throws ClassNotFoundException 
	{
		Check.notNull(className, "className");
		Check.notNull(superClass, "superClass");
		
		Class<?> c = loader != null ? loader.loadClass(className) : Class.forName(className);
		Check.isSuperclassOf(superClass, c);
		return (Class<? extends T>)c;
	}
	
	
	/**
	 * Returns the class for a given class name or null if the class
	 * cannot be found.
	 * @param className the name of the class
	 * @param superClass a superclass of the class. The method checks that the class 
	 * 		is derived from the superclass. 
	 * @throws IllegalArgumentException if the class is not derived from the given superClass. 
	 */
	public static <T> Class<? extends T> getPotentialClass(String className, Class<T> superClass, ClassLoader loader)  
	{
		try
		{
			return getClass(className, superClass, loader);
		}
		catch(ClassNotFoundException e)
		{
			return null;
		}
	}

	
	/**
	 * Tests if the annotation is an instance of the given annotation class.
	 */
	public static <A extends Annotation> boolean isA(Annotation annotation, Class<A> c)
	{
		return (annotation != null) && (c != null) && c.isAssignableFrom(annotation.annotationType());
	}
	

	/**
	 * Returns the first annotation in a list of annotations which is an instance of the
	 * given annotation class.
	 */
	@SuppressWarnings("unchecked")
	public static <A extends Annotation> A findAnnotation(Annotation[] annotations, Class<A> c)
	{
		for (Annotation a : annotations)
		{
			if (isA(a, c))
				return (A)a;
		}
		return null;
	}

	
	/**
	 * Tests if an object is an instance of a given class.
	 * If yes the object is casted to that class and returned, else null is returned.
	 */
	@SuppressWarnings("unchecked")
	public static <T> T unwrap(Object object, Class<T> targetClass)
	{
		return isA(object, targetClass) ? (T)object : null;
	}
	
	
	/**
	 * Translates a Enum name into an Enum object.
	 * If the Enum name is null or not defined, then the default value is returned. This helps 
	 * to avoid the IllegalArgumentException thrown by Enum.valueOf 
	 */
	public static <T extends Enum<T>> T getEnum(Class<T> enumClass, String name, T defaultValue)
	{
		try
		{
			if (name != null)
				return Enum.valueOf(enumClass, name);
		}
		catch(IllegalArgumentException e)
		{
		}
		return defaultValue;
	}
	

	/**
	 * Returns the package name of the class, with all '.' characters
	 * replaced by '/'.
	 */
	public static String buildResourcePath(Class<?> c)
	{
		return getPackageName(c).replace('.', '/');
	}


	/**
	 * Builds the resource path for the class and appends the file name.
	 * The returned string is suitable to be passed to ClasLoader.getResource().
	 */
	public static String buildResourceFile(Class<?> c, String name)
	{
		return buildResourcePath(c) + '/' + name;
	}
}
