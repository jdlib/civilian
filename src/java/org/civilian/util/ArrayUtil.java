/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.util;


import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.Objects;


/**
 * Helper methods for array handling.
 */
public abstract class ArrayUtil
{
	/**
	 * Creates a new array, with the given item prepended.
	 * @param array an array
	 * @param item an item
	 * @param <T> the element type
	 * @return the new array
	 */
	public static <T> T[] addFirst(T[] array, T item)
	{
		@SuppressWarnings("unchecked")
		T[] result = (T[])Array.newInstance(array.getClass().getComponentType(), array.length + 1);
		System.arraycopy(array, 0, result, 1, array.length);
		result[0] = item;
		return result;
	}

	
	/**
	 * Creates a new array, with the given item appended to the end.
	 * @param array an array
	 * @param item an item
	 * @param <T> the element type
	 * @return the new array
	 */
	public static <T> T[] addLast(T[] array, T item)
	{
		@SuppressWarnings("unchecked")
		T[] result = (T[])Array.newInstance(array.getClass().getComponentType(), array.length + 1);
		System.arraycopy(array, 0, result, 0, array.length);
		result[array.length] = item;
		return result;
	}


	/**
	 * Creates a new array, with the given item appended to the end.
	 * @param array1 the first array
	 * @param array2 the second array
	 * @param <T> the element type
	 * @return the new array
	 */
	public static <T> T[] concat(T[] array1, T[] array2)
	{
		if (array1 == null) 
			return array2;
		else if (array2 == null)
			return array1;
		else
		{
			@SuppressWarnings("unchecked")
			T[] result = (T[])Array.newInstance(array1.getClass().getComponentType(), array1.length + array2.length);
			System.arraycopy(array1, 0, result, 0, array1.length);
			System.arraycopy(array2, 0, result, array1.length, array2.length);
			return result;
		}
	}

	
	/**
	 * Returns a new array, with the first count items removed.
	 * @param array an array
	 * @param count a element count
	 * @param <T> the element type
	 * @return the new array
	 */
	public static <T> T[] removeFirst(T[] array, int count)
	{
		if (array.length < count)
			throw new IllegalArgumentException("array length is " + array.length);
		if (count == 0)
			return array;
		
		@SuppressWarnings("unchecked")
		T[] result = (T[])Array.newInstance(array.getClass().getComponentType(), array.length - count);
		System.arraycopy(array, count, result, 0, result.length);
		return result;
	}
	
	
	/**
	 * Returns a new array, with the item at the given index removed
	 * @param <T> the element type
	 * @param array an array
	 * @param index an index into the array
	 * @return the array
	 */
	public static <T> T[] removeAt(T[] array, int index)
	{
		if ((index < 0) || (index >= array.length))
			return array;
		
		@SuppressWarnings("unchecked")
		T[] result = (T[])Array.newInstance(array.getClass().getComponentType(), array.length - 1);
		if (index > 0)
			System.arraycopy(array, 0, result, 0, index);
		if (index < array.length - 1)
			System.arraycopy(array, index + 1, result, index, array.length - 1 - index);
		return result;
	}

	
	/**
	 * Tests if the array contains the item.
	 * @param <T> the element type
	 * @param array an array
	 * @param item an item
	 * @return contains?
	 */
	public static <T> boolean contains(T[] array, T item)
	{
		if (array != null)
		{
			for (int i=0; i<array.length; i++)
			{
				if (Objects.equals(array[i], item))
					return true;
			}
		}
		return false;
	}


	/**
	 * Collects the items of an Enumeration and returns them as an array.
	 * @param it an Enumeration
	 * @param c the element class
	 * @param <T> the element type
	 * @return the array
	 */
	@SuppressWarnings("unchecked")
	public static <T> T[] toArray(Enumeration<T> it, Class<T> c)
	{
		ArrayList<T> list = new ArrayList<>(); 
		while(it.hasMoreElements())
			list.add(it.nextElement());
		return list.toArray((T[])Array.newInstance(c, list.size()));
	}
}
