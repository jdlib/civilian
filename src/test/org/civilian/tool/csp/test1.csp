package org.civilian.template.csp


import java.util.ArrayList


prolog /* Class-Comment */
template(String... names, ArrayList < String,String > someList1, ArrayList<? extends X<String>> someList2) 
	mixin lang, html:htm, html, com.app.MyMixin
{{
	<start>
	@for (int i=0; i<10; i++)
		^?{i % 2 == 0}|conditional|
		^names: ^{lang.format(i)}
		@list_.add(names);
	</start>
}}


private ArrayList<String> list_ = new ArrayList<>();