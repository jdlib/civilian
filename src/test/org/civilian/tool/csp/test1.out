/**
 * Generated from test1.csp
 * Do not edit.
 */
package org.civilian.template.csp;


import java.util.ArrayList;
import com.app.MyMixin;
import org.civilian.template.CspTemplate;
import org.civilian.template.mixin.HtmlMixin;
import org.civilian.template.mixin.LangMixin;


/* Class-Comment */
public class Test1 extends CspTemplate
{
	public Test1(String... names, ArrayList<String,String > someList1, ArrayList<? extends X<String>> someList2)
	{
		this.names = names;
		this.someList1 = someList1;
		this.someList2 = someList2;
	}


	@Override protected void init()
	{
		super.init();
		lang = new LangMixin(out);
		htm = new HtmlMixin(out);
		html = new HtmlMixin(out);
		myMixin = new MyMixin(out);
	}


	@Override protected void exit()
	{
		lang = null;
		htm = null;
		html = null;
		myMixin = null;
		super.exit();
	}


	@Override protected void print() throws Exception
	{
		out.println("<start>");                                         // line 11: <start>
		for (int i=0; i<10; i++)                                        // line 12: @for (int i=0; i<10; i++)
		{
			if (i % 2 == 0)                                             // line 13: ^?{i % 2 == 0}
			{
				out.print("conditional");                               // line 13: conditional
			}
			out.printlnIfNotEmpty();
			out.print(names);                                           // line 14: ^names
			out.print(": ");                                            // line 14: :
			out.print(lang.format(i));                                  // line 14: ^{lang.format(i)}
			out.printlnIfNotEmpty();
			list_.add(names);                                           // line 15: @list_.add(names);
		}
		out.println("</start>");                                        // line 16: </start>
	}
	
	
	private ArrayList<String> list_ = new ArrayList<>();


	protected String[] names;
	protected ArrayList<String,String > someList1;
	protected ArrayList<? extends X<String>> someList2;
	protected LangMixin lang;
	protected HtmlMixin htm;
	protected HtmlMixin html;
	protected MyMixin myMixin;
}
